WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift

```diff
@@ -118,26 +128,82 @@ final class UpgradesViewModel: ObservableObject {
+            /// before we try to advance to the waiting state.
+            DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
+                guard let self else { return }
+                /// If the user cancelled, the state will be `.loaded(_)` by now, so we don't advance to waiting.
+                /// Likewise, errors will have moved us to `.error(_)`, so we won't advance then either.
+                if case .purchasing(_) = self.upgradeViewState {
+                    self.upgradeViewState = .waiting(wooWPComPlan)
+                }
+            }
+        }
+
         do {
             let result = try await inAppPurchasesPlanManager.purchasePlan(with: planID,
                                                                           for: siteID)
-            // TODO: handle `pending` here... somehow – requires research
-            // TODO: handle `.success(.unverified(_))` here... somehow
-            guard case .success(.verified(_)) = result else {
+            stopObservingInAppPurchaseDrawerDismissal()
+            switch result {
+            case .userCancelled:
+                upgradeViewState = .loaded(wooWPComPlan)
+            case .success(.verified(_)):
+                upgradeViewState = .completed
+            default:
+                // TODO: handle `pending` here... somehow – requires research
+                // TODO: handle `.success(.unverified(_))` here... somehow
                 return
             }
-            upgradeViewState = .completed
         } catch {
             DDLogError("purchasePlan \(error)")
+            stopObservingInAppPurchaseDrawerDismissal()
             upgradeViewState = .error(UpgradesError.purchaseError)
         }
     }
 
+    /// Observes the `didBecomeActiveNotification` for one invocation of the notification.
+    /// Using this in the scope of `purchasePlan` tells us when Apple's IAP view has completed.
+    ///
+    /// However, it can also be triggered by other actions, e.g. a phone call ending.
+    ///
+    /// One good example test is to start an IAP, then background the app and foreground it again
+    /// before the IAP drawer is shown.  You'll see that this notification is received, even though the
+    /// IAP drawer is then shown on top. Dismissing or completing the IAP will not then trigger this
+    /// notification again.
+    ///
+    /// It's not perfect, but it's what we have.
+    private func observeInAppPurchaseDrawerDismissal(whenFired action: @escaping (() -> Void)) {
+        applicationDidBecomeActiveObservationToken = notificationCenter.addObserver(
+            forName: UIApplication.didBecomeActiveNotification,
+            object: nil,
+            queue: .main) { [weak self] notification in
+                action()
+                self?.stopObservingInAppPurchaseDrawerDismissal()
+            }
+    }
+
+    private func stopObservingInAppPurchaseDrawerDismissal() {
```

Comment:
Should we add an extra call to stop observing for notifications inside a deinit method as well? While it should stop when the `UpgradesViewModel` is deallocated, I wonder if we could face some memory leak/unexpected behaviour issues due to the multiple async tasks and/or potential strong references:
```
deinit {
    stopObservingInAppPurchaseDrawerDismissal()
}
```